dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromPrompt.cs

```diff
@@ -228,5 +245,59 @@ private async Task<(ITextGenerationService, string, PromptRenderedEventArgs?)> R
     /// <summary>Create a random, valid function name.</summary>
     private static string RandomFunctionName() => $"func{Guid.NewGuid():N}";
 
+    /// <summary>
+    /// Captures usage details, including token information.
+    /// </summary>
+    private void CaptureUsageDetails(string? modelId, IDictionary<string, object?>? metadata, ILogger logger)
+    {
+        if (string.IsNullOrWhiteSpace(modelId))
+        {
+            logger.LogWarning("No model ID provided to capture usage details.");
+            return;
+        }
+
+        if (metadata is null)
+        {
+            logger.LogWarning("No metadata provided to capture usage details.");
+            return;
+        }
+
+        if (!metadata.TryGetValue("Usage", out object? usageObject) || usageObject is null)
+        {
+            logger.LogWarning("No usage details provided to capture usage details.");
+            return;
+        }
+
+        var promptTokens = 0;
+        var completionTokens = 0;
+        try
+        {
+            var jsonObject = JsonSerializer.Deserialize<JsonElement>(JsonSerializer.Serialize(usageObject));
+            promptTokens = jsonObject.GetProperty("PromptTokens").GetInt32();
+            completionTokens = jsonObject.GetProperty("CompletionTokens").GetInt32();
```

Comment:
Can we avoid roundtripping through JSON to get at this information? It adds a lot of overhead. And relying on exceptions if it can't parse makes it super expensive if we're getting back different objects.

Reply:
Unfortunately, the way the token information is returned would require us to perform the round trip. The object that contains the token usage is of type `Azure.AI.OpenAI.CompletionsUsage`. We don't want to take a dependency on OpenAI at the function level, thus we have to rely on Json.

Reply:
I have changed it to use `SerializeToElement`, which simplifies the code but I believe it performs the same operation underneath.